e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -219,3 +225,314 @@ def sample_params(self) -> Dict[str, int]:
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(request_id=create.request_id, length=self.rungs[0].units_needed)
+            )
+
+            self.trial_rungs[create.request_id] = 0
+            self.pending_trials += 1
+
+            return ops
+
+        self.early_exit_trials[request_id] = True
+        self.closed_trials[request_id] = True
+        return self.promote_async(request_id, 100.0)
+
+    def initial_operations(self) -> List[Operation]:
+        ops: List[Operation] = []
+
+        if self.max_concurrent_trials > 0:
+            max_concurrent_trials = min(self.max_concurrent_trials, self.max_trials)
+        else:
+            max_concurrent_trials = max(
+                1, min(int(pow(self.divisor, self.num_rungs - 1)), self.max_trials)
+            )
+
+        for _ in range(0, max_concurrent_trials):
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(request_id=create.request_id, length=self.rungs[0].units_needed)
+            )
+
+            self.trial_rungs[create.request_id] = 0
+            self.pending_trials += 1
+
+        return ops
+
+    def promote_async(self, request_id: uuid.UUID, metric: float) -> List[Operation]:
+        rung_idx = self.trial_rungs[request_id]
+        rung = self.rungs[rung_idx]
+        rung.outstanding_trials -= 1
+        added_train_workload = False
+
+        ops: List[Operation] = []
+
+        if rung_idx == self.num_rungs - 1:
+            rung.metrics.append(TrialMetric(request_id=request_id, metric=metric))
+
+            if request_id in self.early_exit_trials and self.early_exit_trials[request_id] is True:
+                ops.append(Close(request_id=request_id))
+                logging.info(f"Closing trial {request_id}")
+                self.closed_trials[request_id] = True
+        else:
+            next_rung = self.rungs[rung_idx + 1]
+            logging.info(f"Promoting in rung {rung_idx}")
+            for promoted_request_id in rung.promotions_async(request_id, metric, self.divisor):
+                self.trial_rungs[promoted_request_id] = rung_idx + 1
+                next_rung.outstanding_trials += 1
+                if (
+                    request_id not in self.early_exit_trials
+                    or not self.early_exit_trials[promoted_request_id]
+                ):
+                    logging.info(f"Promoted {promoted_request_id}")
+                    units_needed = max(next_rung.units_needed - rung.units_needed, 1)
+                    ops.append(ValidateAfter(promoted_request_id, units_needed))
+                    added_train_workload = True
+                    self.pending_trials += 1
+                else:
+                    return self.promote_async(promoted_request_id, 100.0)
```

Comment:
Should we use a number larger than 100.0 to be the equivalent of `math.MaxFloat64`? Maybe `sys.float_info.max` would do the trick?

Reply:
My bad. Should be max float.